File: hibernate/src/main/java/io/github/robertomike/constructors/ConstructJoinFetch.java
package io.github.robertomike.constructors;

import io.github.robertomike.actions.JoinFetch;
import jakarta.persistence.criteria.Root;

public class ConstructJoinFetch extends Construct<JoinFetch> {
    public void construct(Root<?> root) {
        items.forEach(value -> root.fetch(value.getRelationship(), value.getJoinType()));
    }
}

File: hibernate-query-language/src/main/kotlin/io/github/robertomike/hefesto/hql/constructors/ConstructJoinFetch.kt
package io.github.robertomike.hefesto.hql.constructors

import io.github.robertomike.hefesto.actions.JoinFetch
import io.github.robertomike.hefesto.hql.builders.Hefesto
import io.github.robertomike.hefesto.constructors.Construct

class ConstructJoinFetch : Construct<JoinFetch>() {
    fun construct(builder: Hefesto<*>): String {
        val joinQuery = StringBuilder()

        items.forEach { value ->
            if (joinQuery.isNotEmpty()) {
                joinQuery.append(" ")
            }

            joinQuery.append(apply(builder, value))
        }

        return joinQuery.toString()
    }

    fun apply(builder: Hefesto<*>, join: JoinFetch): String {
        return "${join.joinType} join fetch ${if (join.nested) "" else "${builder.acronymTable}."}${join.relationship}${if (join.alias != null) " as ${join.alias}" else ""}"
    }
}
